API: SubdatasetInfo

elpaso opened the pull request:
Fix #7261 

Initial GDALSubdatasetInfo API.

elpaso commented:
@rouault I'd like a general review of the API before I get through all the formats, for now the only implemented one is GPKG.

I will also probably need a list of the drivers that support subdatasets.

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+ * @brief Returns a new GDALSubdatasetInfo object with methods to extract
+ *        and manipulate subdataset information.
+ *        If the pszFileName argument is not recognized by any driver as
+ *        a subdataset descriptor NULL is returned.
```
```suggestion
 *        a subdataset descriptor, NULL is returned.
```

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+GDALGetSubdatasetInfo(const char *pszFileName);
+
+/**
+ * @brief GDALSubdatasetInfoGetFileName returns the file path (or base) component of a
```
Conventionally, the name of the function isn't contained in the brief

```suggestion
 * @brief Returns the file path (or base) component of a
```

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+    GDALSubdatasetInfoH hInfo, const char *pszFileName);
+
+/**
+ * @brief GDALSubdatasetInfoModifyFileName replaces the base component of a
```
```suggestion
 * @brief Replaces the base component of a
```

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+    const char *pszNewFileName);
+
+/**
+ * @brief GDALDestroySubdatasetInfo destroys a GDALSubdatasetInfo object.
```
```suggestion
 * @brief Destroys a GDALSubdatasetInfo object.
```

rouault commented on gcore/gdal_priv.h:
```diff
@@ -60,6 +60,7 @@ class GDALRelationship;
 
 #include "gdal.h"
 #include "gdal_frmts.h"
+#include "gdalsubdatasetinfo.h"
```
this header file should be added in target_public_header() in gcore/CMakeLists.txt

rouault commented on gcore/gdalsubdatasetinfo.cpp:
```diff
@@ -0,0 +1,96 @@
+ ---------------------
+ begin                : 21.7.2023
+ copyright            : (C) 2023 by ale
+ email                : [your-email-here]
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
```
to be replaced by the MIT license

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+ * @return                      Opaque pointer to a GDALSubdatasetInfo object or NULL if no drivers accepted the file name.
+ * @since                       GDAL 3.8
+ */
+GDALSubdatasetInfoH CPL_DLL CPL_STDCALL
```
Just raising this once, but the CPL_STDCALL are mostly a historical remain of the times of VB6 if I remember well. So they could be removed in the .h and .cpp

elpaso replied:
If that's the case they should be removed from all source files not just these ones.

rouault replied:
Changing the call convention is an ABI break, so we cannot do that lightly. My point was that there was no need to perpetuate using CPL_STDCALL for new API.

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+ * @return                      The original string with the subdataset information removed.
+ * @since                       GDAL 3.8
+ */
+const char CPL_DLL *CPL_STDCALL GDALSubdatasetInfoGetFileName(
```
"const char*" is for returns of non-freeable strings

```suggestion
char CPL_DLL *CPL_STDCALL GDALSubdatasetInfoGetFileName(
```

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,79 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+ * @return                      The original string with the old file name replaced by newFileName and the subdataset information unaltered.
+ * @since                       GDAL 3.8
+ */
+const char CPL_DLL *CPL_STDCALL GDALSubdatasetInfoModifyFileName(
```
```suggestion
char CPL_DLL *CPL_STDCALL GDALSubdatasetInfoModifyFileName(
```

rouault commented on gcore/gdalsubdatasetinfo.cpp:
```diff
@@ -0,0 +1,96 @@
+const char *CPL_STDCALL GDALSubdatasetInfoGetFileName(GDALSubdatasetInfoH hInfo,
+                                                      const char *pszFileName)
+{
+    const std::string fileName{
+        hInfo->GetFilenameFromSubdatasetName(pszFileName)};
+    char *ptr = new char[fileName.size() + 1];
+    strcpy(ptr, fileName.c_str());
+    return ptr;
```
```suggestion
    return CPLStrdup(hInfo->GetFilenameFromSubdatasetName(pszFileName).c_str());
```

rouault commented on gcore/gdalsubdatasetinfo.cpp:
```diff
@@ -0,0 +1,96 @@
+    GDALSubdatasetInfoH hInfo, const char *pszFileName,
+    const char *pszNewFileName)
+{
+    const std::string fileName{
+        hInfo->ModifyFileName(pszFileName, pszNewFileName)};
+    char *ptr = new char[fileName.size() + 1];
+    strcpy(ptr, fileName.c_str());
+    return ptr;
```
```suggestion
    return CPLStrdup(hInfo->ModifyFileName(pszFileName, pszNewFileName).c_str());
```

rouault commented on gcore/gdalsubdatasetinfo.h:
```diff
@@ -0,0 +1,80 @@
+ begin                : 21.7.2023
+ copyright            : (C) 2023 by ale
+ email                : [your-email-here]
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
```
MIT license

rouault commented on gcore/gdal_priv.h:
```diff
@@ -1812,6 +1813,9 @@ class CPL_DLL GDALDriver : public GDALMajorObject
         GDALProgressFunc pfnProgress, void *pProgressData) = nullptr;
 
+    /* Returns a (possibly null) pointer to the Subdataset informational function */
+    GDALSubdatasetInfo *(*pfnGetSubdatasetInfoFunc)() = nullptr;
```
I would have make pfnGetSubdatasetInfoFunc take a const char* argument with the subdataset string. That way this saves the GDALGetSubdatasetInfo::IsSubdatasetSyntax() method, and passing the subdataset name as an argument of each method of GDALGetSubdatasetInfo
(even if that method was kept, I don't think it should be exposed in the C API, as it is mostly an implementation detail of GDALSubdatasetInfo)

rouault commented on ogr/ogrsf_frmts/gpkg/ogrgeopackagedriver.cpp:
```diff
@@ -246,6 +246,88 @@ static int OGRGeoPackageDriverIdentify(GDALOpenInfo *poOpenInfo)
+{
+    virtual ~OGRGeoPackageDriverSubdatasetInfo() = default;
+
+    bool IsSubdatasetSyntax(const std::string &fileName) const override
```
seeing this implementation indeed makes me thinks that my above remark about pfnGetSubdatasetInfoFunc taking the subdataset name to be probably better, as saving lines of code. The components of the subdataset could be saved as member variable(s) of the class during construction

The advantage of your approach is that it enables to deal with several subdatasets with a same GDALSubdatasetInfo instance, but I don't think that's a real use case. And the name SubdatasetInfo would make me think that this is a class whose instances know about a particular subdataset

elpaso replied:
Sounds good, I had the same doubt but reading the issue comments it wasn't totally clear which way you wanted to go.

rouault commented on ogr/ogrsf_frmts/gpkg/ogrgeopackagedriver.cpp:
```diff
@@ -246,6 +246,88 @@ static int OGRGeoPackageDriverIdentify(GDALOpenInfo *poOpenInfo)
+            return false;
+        }
+
+        // Split
+        char **papszParts{CSLTokenizeString2(fileName.c_str(), ":", 0)};
+        const int iPartsCount{CSLCount(papszParts)};
+        CSLDestroy(papszParts);
+        return iPartsCount > 2;
```
This test is too generic. It would catch "c:\foo\bar.gpkg" as a subdataset name

rouault commented on ogr/ogrsf_frmts/gpkg/ogrgeopackagedriver.cpp:
```diff
@@ -354,6 +436,8 @@ class GDALGPKGDriver final : public GDALDriver
         InitializeCreationOptionList();
         return GDALDriver::GetMetadata(pszDomain);
     }
+
+    GDALSubdatasetInfo *GetSubdatasetInfo();
```
Probably a remain of initial tries:

```suggestion
```

rouault commented on swig/include/SubdatasetInfo.i:
```diff
@@ -0,0 +1,89 @@
+
+%extend {
+
+        GDALSubdatasetInfoShadow(){
+            return new GDALSubdatasetInfo();
+        }
+
```
We don't need a constructor:

```suggestion
```

elpaso replied:
I doesn't build without it.

rouault replied:
looking at other SWIG classes, you need to use that pattern:
```
class GDALSubdatasetInfoShadow {
private:
  GDALSubdatasetInfoShadow();
public:
%extend {
```

rouault commented on swig/include/SubdatasetInfo.i:
```diff
@@ -0,0 +1,89 @@
+            GDALDestroySubdatasetInfo(reinterpret_cast<GDALSubdatasetInfoH>(self));
+        }
+
+        const char* GetFilenameFromSubdatasetName(const char *pszFileName)
```
```suggestion
        retStringAndCPLFree* GetFilenameFromSubdatasetName(const char *pszFileName)
```

rouault commented on swig/include/SubdatasetInfo.i:
```diff
@@ -0,0 +1,89 @@
+            return GDALSubdatasetInfoGetFileName(reinterpret_cast<GDALSubdatasetInfoH>(self), pszFileName );
+        }
+
+        const char* ModifyFileName(const char *pszFileName, const char *pszNewFileName)
```
```suggestion
        retStringAndCPLFree* ModifyFileName(const char *pszFileName, const char *pszNewFileName)
```

rouault commented on swig/include/SubdatasetInfo.i:
```diff
@@ -0,0 +1,89 @@
+}
+};
+
+%newobject GetSubdatasetInfo(const char *pszFileName);
```
was probably fine, but it seems we use the short form usually:
```suggestion
%newobject GetSubdatasetInfo;
```

rouault commented on gcore/gdalsubdatasetinfo.cpp:
```diff
@@ -0,0 +1,137 @@
+/*                     Subdataset informational functions               */
+/************************************************************************/
+
+GDALSubdatasetInfoH CPL_STDCALL GDALGetSubdatasetInfo(const char *pszFileName)
```
```suggestion
GDALSubdatasetInfoH GDALGetSubdatasetInfo(const char *pszFileName)
```

rouault commented on gcore/gdalsubdatasetinfo.h:
```diff
@@ -0,0 +1,94 @@
+     */
+    virtual void parseFileName() = 0;
+    mutable bool m_initialized = false;
+    std::string m_fileName;
+    std::string m_baseComponent;
```
The difference of semantics between those 2 is unclear to me.
I believe some comments showing how an example subdataset name is split into all those member variables would be helpful

elpaso replied:
m_fileName is the unaltered string set in the constructor, the *Component are the individual pieces of information after the parsing (split `:`) operation

rouault commented on gcore/gdalsubdatasetinfo.cpp:
```diff
@@ -0,0 +1,137 @@
+        this_->parseFileName();
+        m_initialized = true;
+    }
+    return m_driverPrefixComponent + ":" + m_baseComponent;
```
I would expect this to be just:
```suggestion
    return m_filename;
```
I would avoid assuming at this generic level that a ':' is used to join m_driverPrefixComponent and m_baseComponent
I guess if there is a need to return m_driverPrefixComponent + ":" + m_baseComponent that should be through a virtual method implemented in the driver

elpaso replied:
Ok, my idea was to return the path to the parent dataset with the subdataset information stripped off, I thought it was useful but we can add it later if needed.

rouault commented on gcore/gdal.h:
```diff
@@ -1290,6 +1290,71 @@ bool CPL_DLL GDALDatasetSetQueryLoggerFunc(
+ * @return                      The subdataset name.
+ * @since                       GDAL 3.8
+ */
+char CPL_DLL *GDALSubdatasetInfoGetSubdatasetName(GDALSubdatasetInfoH hInfo);
```
I would rather perhaps call it
```suggestion
char CPL_DLL *GDALSubdatasetInfoGetSubdatasetComponent(GDALSubdatasetInfoH hInfo);
```
(and with related changes at the C++ and SWIG level)

since in the terminology used else where the subdataset name is the full  "netCDF:foo.nc:my_variable"

elpaso commented:
@rouault I agree there is a little confusion in the terminology, here is my proposal to clarify and avoid ambiguity:

The complete file name (e.g. `GPKG:/path_to.gpkg:layer_name:var_name`) will be named "File name"  and the methods and variables will be renamed accordingly.

The parsed components will be named:

- driver prefix e.g. `GPKG`  -> "Prefix component"
- file path: e.g. `/path_to.gpkg` -> "Path component"
- subdataset: e.g. `layer_name` -> "Subdataset component"

So, the methods will become:

GetFileName()  -> returns the original complete file name e.g. `GPKG:/path_to.gpkg:layer_name:var_name`
GetPathComponent() ->  returns the path component e.g. `/path_to.gpkg` 
GetPrefixComponent() -> returns the prefix component e.g. `GPKG` 
GetSubdatasetComponent() -> returns the subdataset component  e.g. `layer_name`
ModifyPathComponent( newPathComponent) -> replaces the path component

Or, the shorter versions without `Component`:
GetFileName() 
GetPath()
GetPrefix()
GetSubdataset()
ModifyPath(newPath)

rouault commented:
> GetSubdataset()

or maybe GetSuffix() ?

elpaso commented:
> > GetSubdataset()
> 
> or maybe GetSuffix() ?

I don't follow: isn't subdataset more appropriate to describe it? Suffix is really generic and if we have variables it might not even be the last component in the descriptor.

rouault commented:
> I don't follow: isn't subdataset more appropriate to describe it? Suffix is really generic and if we have variables it might not even be the last component in the descriptor.

ok, let's keep GetSubdataset(). "There are only two hard things in Computer Science: cache invalidation and naming things. " :-)

rouault commented on frmts/gtiff/geotiff.cpp:
```diff
@@ -1132,6 +1132,84 @@ CPLString GTiffGetCompressValues(bool &bHasLZW, bool &bHasDEFLATE,
+/*                    OGRGTiffDriverGetSubdatasetInfo()                 */
+/************************************************************************/
+
+struct OGRGTiffDriverSubdatasetInfo : public GDALSubdatasetInfo
```
should rather be:
```suggestion
struct GDALGTiffDriverSubdatasetInfo : public GDALSubdatasetInfo
```
or maybe just:
```suggestion
struct GTiffDriverSubdatasetInfo : public GDALSubdatasetInfo
```

rouault commented on frmts/gtiff/geotiff.cpp:
```diff
@@ -1132,6 +1132,84 @@ CPLString GTiffGetCompressValues(bool &bHasLZW, bool &bHasDEFLATE,
+            return;
+        }
+
+        char **papszParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
```
you can use CPLStringList to avoid having to explicitly free
```suggestion
        CPLStringList aosParts(CSLTokenizeString2(m_fileName.c_str(), ":", 0));
```

rouault commented on frmts/gtiff/geotiff.cpp:
```diff
@@ -1132,6 +1132,84 @@ CPLString GTiffGetCompressValues(bool &bHasLZW, bool &bHasDEFLATE,
+
+    // GDALSubdatasetInfo interface
+  private:
+    void parseFileName() override
```
what do you think about the opportunity of using this in the GTiffDataset::OpenDir() method to avoid duplication of parsing ? (this could apply to other drivers as well)

elpaso replied:
I'll have a look, I prefer a method that does not have to open the file though, ideally this API should just manipulate strings without I/O overhead.

rouault commented on frmts/gtiff/geotiff.cpp:
```diff
@@ -1132,6 +1132,84 @@ CPLString GTiffGetCompressValues(bool &bHasLZW, bool &bHasDEFLATE,
+        {
+            return;
+        }
+
```
There is a more esoteric supported variant:   GTIFF_DIR:off:<dir_offset>:filename
If we reused this class for GTiffDataset::OpenDir(), we'd have to support it

rouault commented on frmts/postgisraster/postgisrasterdataset.cpp:
```diff
@@ -4109,6 +4109,86 @@ GBool PostGISRasterDataset::PolygonFromCoords(int nXOff, int nYOff,
+/*                    OGRPostGISRasterDriverGetSubdatasetInfo()            */
+/************************************************************************/
+
+struct OGRPostGISRasterDriverSubdatasetInfo : public GDALSubdatasetInfo
```
```suggestion
struct PostGISRasterDriverSubdatasetInfo : public GDALSubdatasetInfo
```

rouault commented on frmts/wms/wmsdriver.cpp:
```diff
@@ -1132,6 +1132,67 @@ void WMSDeregister(CPL_UNUSED GDALDriver *d)
+/*                    OGRWMSDriverGetSubdatasetInfo()                   */
+/************************************************************************/
+
+struct OGRWMSDriverSubdatasetInfo : public GDALSubdatasetInfo
```
```suggestion
struct WMSDriverSubdatasetInfo : public GDALSubdatasetInfo
```

rouault commented on ogr/ogrsf_frmts/gpkg/ogrgeopackagedriver.cpp:
```diff
@@ -246,6 +246,86 @@ static int OGRGeoPackageDriverIdentify(GDALOpenInfo *poOpenInfo)
+            return;
+        }
+
+        char **papszParts{CSLTokenizeString2(m_fileName.c_str(), ":", 0)};
```
```suggestion
        CPLStringList aosParts(CSLTokenizeString2(m_fileName.c_str(), ":", 0));
```

rouault commented on frmts/gtiff/geotiff.cpp:
```diff
@@ -1132,6 +1132,82 @@ CPLString GTiffGetCompressValues(bool &bHasLZW, bool &bHasDEFLATE,
+                // Invalid
+                if (!hasDriveLetter)
+                {
+                    CSLDestroy(aosParts);
```
```suggestion
```

rouault commented on frmts/gtiff/geotiff.cpp:
```diff
@@ -1132,6 +1132,82 @@ CPLString GTiffGetCompressValues(bool &bHasLZW, bool &bHasDEFLATE,
+            {
+                if (hasDriveLetter)
+                {
+                    CSLDestroy(aosParts);
```
```suggestion
```

rouault commented on ogr/ogrsf_frmts/gpkg/ogrgeopackagedriver.cpp:
```diff
@@ -246,6 +246,85 @@ static int OGRGeoPackageDriverIdentify(GDALOpenInfo *poOpenInfo)
+                // Invalid
+                if (!hasDriveLetter)
+                {
+                    CSLDestroy(aosParts);
```
```suggestion
```

rouault commented on ogr/ogrsf_frmts/gpkg/ogrgeopackagedriver.cpp:
```diff
@@ -246,6 +246,85 @@ static int OGRGeoPackageDriverIdentify(GDALOpenInfo *poOpenInfo)
+            {
+                if (hasDriveLetter)
+                {
+                    CSLDestroy(aosParts);
```
```suggestion
```

sgillies commented:
@rouault @elpaso how did we arrive at using `:` as a delimiter between dataset and subdataset name? Does this have to be extended to drivers where it doesn't exist already? I think `:` is a poor choice and leads to people putting quotes around dataset URLs within a GDAL filename.

elpaso commented:
> @rouault @elpaso how did we arrive at using `:` as a delimiter between dataset and subdataset name? Does this have to be extended to drivers where it doesn't exist already? I think `:` is a poor choice and leads to people putting quotes around dataset URLs within a GDAL filename.

I don't know where the colon separator comes from, but I totally agree that it's a poor choice, the QGIS counterpart QgsDatasourceUri is a little bit better engineered, for most datasources it uses URL encoding for the KVP and it allows to encode the opening options.

rouault commented:
> @rouault @elpaso how did we arrive at using `:` as a delimiter between dataset and subdataset name? Does this have to be extended to drivers where it doesn't exist already? I think `:` is a poor choice and leads to people putting quotes around dataset URLs within a GDAL filename.

@sgillies  Before my involvement with the project. I can see that this syntax was already used in the initial version of the HDF4 driver back in 2002:  [7c56e20797c17d2bcb3f173d652ec1fdf590d558](https://github.com/OSGeo/gdal/commit/7c56e20797c17d2bcb3f173d652ec1fdf590d558#diff-5b0f8cff7aa04486157ad10213a9a297a72afa1301576a3071f3352f86c0ef13R656)
While this is indeed not ideal, changing the subdataset syntax in existing drivers would have backward compatibility issues. I suspect a number of people have VRT files referencing the old syntax, or some code manually composing them.
The new API itself in this PR makes no assumption on the separator, so I don't think it is a step backwards, and it should be hopefully compatible with better subsdataset syntax.
That said the replacement logic in ModifyPathComponent() is fully string based currently. if at some point a driver would URI-encode the filename, that wouldn't work. But if that occured, the method could be made virtual and implemented by the driver that would deal with the URI encoding/decoding.

@elpaso  I believe to keep this PR manageable we should target the most popular file-based drivers that use subsdataset (others might follow in subsequent PR): netCDF, HDF4, HDF5, GPKG . Anything else ?
I'd be interested however to see frmts/vrt/vrtsources.cpp updated to remove the NETCDF releated hacks in favor of the new API to demonstrate its usefulness (hum I see that to remove all of them we should also tackle NITF, PDF, RASTERLITE and TILEDB)

elpaso commented:
> > @rouault @elpaso how did we arrive at using `:` as a delimiter between dataset and subdataset name? Does this have to be extended to drivers where it doesn't exist already? I think `:` is a poor choice and leads to people putting quotes around dataset URLs within a GDAL filename.
> 
> @sgillies Before my involvement with the project. I can see that this syntax was already used in the initial version of the HDF4 driver back in 2002: [7c56e20797c17d2bcb3f173d652ec1fdf590d558](https://github.com/OSGeo/gdal/commit/7c56e20797c17d2bcb3f173d652ec1fdf590d558#diff-5b0f8cff7aa04486157ad10213a9a297a72afa1301576a3071f3352f86c0ef13R656) While this is indeed not ideal, changing the subdataset syntax in existing drivers would have backward compatibility issues. I suspect a number of people have VRT files referencing the old syntax, or some code manually composing them. The new API itself in this PR makes no assumption on the separator, so I don't think it is a step backwards, and it should be hopefully compatible with better subsdataset syntax. That said the replacement logic in ModifyPathComponent() is fully string based currently. if at some point a driver would URI-encode the filename, that wouldn't work. But if that occured, the method could be made virtual and implemented by the driver that would deal with the URI encoding/decoding.
> 

Yeah.

> @elpaso I believe to keep this PR manageable we should target the most popular file-based drivers that use subsdataset (others might follow in subsequent PR): netCDF, HDF4, HDF5, GPKG . Anything else ? I'd be interested however to see frmts/vrt/vrtsources.cpp updated to remove the NETCDF releated hacks in favor of the new API to demonstrate its usefulness (hum I see that to remove all of them we should also tackle NITF, PDF, RASTERLITE and TILEDB)

I have actually already implemented WMS and PostGIS raster because I wanted to test the API with a DB and a webservice. I will proceed according to your priority list.

elpaso commented:
@rouault here is another iteration. I have removed the double quotes from the logic by storing both the original unmodified path (used only internally to modify the path) and I am storing isQuoted to determine if the new path should be quoted (if it is already quoted it is not quoted twice).

When returning the path is is always unquoted and unescaped.

I have added a few new test cases.

rouault commented:
@sgillies Do you think we can merge this PR or do you want the RFC you mention in https://lists.osgeo.org/pipermail/gdal-dev/2023-September/057641.html to be a pre-condition for it ?

rouault commented:
@elpaso Can you rebase on top of latest master and fix conflict in autotest/gdrivers/netcdf.py?

elpaso commented:
> @elpaso Can you rebase on top of latest master and fix conflict in autotest/gdrivers/netcdf.py?

done

rouault commented on autotest/gdrivers/netcdf.py:
```diff
@@ -6451,3 +6451,73 @@ def test_netcdf_NASA_EMIT():
+###########################################################
+# Test gdal subdataset informational functions
+
+@pytest.mark.parametrize(
```
```suggestion

@pytest.mark.parametrize(
```
